### ISSUE TITLE
Create compilation directives preprocessor

### ISSUE THREAD

**Issue by autoboosh** [rubberduck-vba/Rubberduck#1132]
The current grammar contains compiler directives which makes certain things more complicated, like allowing multiple declarations to exist. How about we run a preprocessor on the source code and then use that as input to the normal VBA parser? We'd have to implement a small interpreter but that might be possible since conditional compilation constants are quite constrained.

Useful links:

1. [Using Conditional Compilation](https://msdn.microsoft.com/en-us/library/aa240847(v=vs.60).aspx)
2. [Conditional Compilation Expressions](https://msdn.microsoft.com/en-us/library/ee200188.aspx)
3. [Conditional Compilation](https://msdn.microsoft.com/en-us/library/ee199757.aspx)

Some things that seemed useful to note:

1. Project property constants are visible project wide. Module constants are private to the module.
2. Constants can refer to other constants
3. The `Like` operator is supported, i.e. we have to interpret the `Like `mini-language as well

Thoughts? :)

**Comment by Vogel612** [comment]
What about refactorings like Rename? We'd have to keep these in mind when compiling the directives. In essence we'd sacrifice Refactorings for Simplicity.

I'd only do that if I had no other choice. Then again you could try to create parseable code for every possible combination of compiler directives and "fix" Refactorings that way. It'd be really hard though.

Things like "Move Declaration Closer to Usage" would become largely impractical...

The list of problematic Refactorings probably goes on, those are just off the top of my head. :neutral_face: If we do this it should be clear what we sacrifice where

**Comment by autoboosh** [comment]
How do tools like R# and Visual Studio deal with this?

**Comment by retailcoder** [comment]
@autoboosh excellent question.

![image](https://cloud.githubusercontent.com/assets/5751684/13205154/f9f04336-d8ae-11e5-9b8a-fae018878556.png)

R# grays out the "dead path". And then renaming `foo` to `bar` leaves the "dead path" as `foo`.

![image](https://cloud.githubusercontent.com/assets/5751684/13205163/4debeda0-d8af-11e5-8bc4-d8bbe4d89d56.png)

In other words, if we can preprocess the input string to, say, replace the "dead path" with whitespaces, we'll have Rubberduck behave exactly as VS/R#.

**Comment by ThunderFrame** [comment]
VBE *will* recognise the dead paths, for Find/Replace, but *won't* enforce syntax on, complile or run the dead paths.

I *think* VBE only populates the code-pane drop-downs (for event navigation, and procedure navigation) for the live-code branch.

IMO, we need to have RubberDuck be *aware* of which code branch a declaration/line belongs to. That way you can offer the user options on the scope of a rename/refactor.

For example, the user might have:

```vb
#Const USE_LEGACY_FORMULA = False

'Block-level directive
#If USE_LEGACY_FORMULA Then
Function Foo() As Integer
  Foo = CInt(Right(Year(Now()),2))
End Function
#Else
Function Foo() As Long
  Foo = Year(Now())
End Function
#End If

'Intra-block directives
Function Bar() As Integer
  #If USE_LEGACY_FORMULA Then
  Bar = Foo()
  #Else
  Bar = Foo()
  #End If
End Function
```

If Rubberduck was aware of the conditional branches, then refactor/renames could be targeted/scoped to include *All* or include/exclude *specific* branches.

In the exanple above, the user could choose to *only* rename `Foo` in the live branch, or only in the USE_LEGACY_FORMULA branch, or in *All* branches.

I've seen projects use Compiler directives in peculiar ways, so while the user *ought* to rename/refactor *Globally*, it doesn't mean that Rubberduck should force them to.

**Comment by retailcoder** [comment]
@ThunderFrame find/replace in the VBE has nothing to do with code paths, it's all about the *text in the editor*, nothing mor: there's nothing to "recognize" there, in fact the VBE actively ignores whatever isn't in the "live" path.

I think we could have an `IsAlive` property on declarations and identifier references, set to `false` when it's located in a "dead branch" - and `true` otherwise. The hard part remains to determine which code is "live", but once that's done we can easily instruct the resolver to only care for `IsAlive` declarations, and things that need to care for "dead" code can.

For example refactor/rename could see that there's a "dead" `Foo` declaration in scope, and *prompt whether to also rename that declaration*.

**Comment by ThunderFrame** [comment]
Indeed, "dead paths, for Find/Replace" ==  "text in the editor"

**Comment by autoboosh** [comment]
Here's a hacked together solution.

It works like this:
1. Generate VBA method that calculates which code paths are "alive"
2. Execute generated method in host application
3. Use result to comment out dead code
4. Pass processed code to normal parser

This code:

```
#Const DEBUG_ = True
#If DEBUG_ Then
    Dim foo As Integer
#Else 
    Dim bar As Integer
#EndIf 
```

Generates the following method:

```
Public Function Liveliness_2f91dcbd11204f7ba4c8dd2e645f3c10() As Boolean()
    Dim livePaths(0 To 1) As Boolean
    Const VBA6 = False
    Const VBA7 = True
    Const Win64 = True
    Const Win32 = True
    Const Win16 = False
    Const Mac = False
    Const DEBUG_ = True
    If DEBUG_ Then
        livePaths(0) = True
    Else
        livePaths(1) = True
    End If
    Liveliness_2f91dcbd11204f7ba4c8dd2e645f3c10 = livePaths
End Function
```

Which when processed leads to the following code:

```
'#Const DEBUG_ = True
'#If DEBUG_ Then
    Dim foo As Integer
'#Else
'    Dim bar As Integer
'#End 
```

Advantages:
1. We don't have to deal with evaluation rules. The host application does this for us.
2. Easily integrated with existing code base / parser

Disadvantages:
1. Only works if compiler directives are used in modules/classes (the run method doesn't return anything otherwise?)
2. Only works with certain host applications (a few host applications apparently do not support return values?)
3. The original source code is slightly altered leading to a source code <-> parse tree mismatch (only affects dead code and conditional compilation nodes though)

I have no idea if this is something we could use, but here it is anyway.

**Comment by autoboosh** [comment]
How is the interpreter going to be integrated into rubberduck? Is e.g. the DeclarationSymbolsListener going to give the interpreter an expression to evaluate and based on the result treat the code as dead/alive? Or should the interpreter use the approach mentioned in the last comment and simply comment out dead code?

**Comment by retailcoder** [comment]
I was thinking of a "preparse" step, involving a dedicated listener/visitor, responsible only for telling which lines to be ignored by the "actual" parser. I'd prefer a non-obstructive solution; commenting-out users' code is less than ideal, it defeats the purpose of precompiler directives.

The listener (visitor?) is given a dictionary of constants and their respective values, adds any `#Const` it finds as the parse tree gets traversed, and then yeah, a visitor needs to be able to interpret the conditionals and determine whether the following block (using a stack to track where it's at) is dead or live - when we determine a block is dead code, we add its start and end line numbers to our result; once we're done, we pick up the results and we can replace the 'dead lines' with whitespace in the input string which we then give to our actual parser.

This preprocessing shouldn't affect performance too much.

**Comment by autoboosh** [comment]
Next try, this time using a hand-built interpreter. Evaluates expressions, does not differentiate all number types though so there might be some edge cases where different results may come up due to rounding errors etc.

The preprocessor turns this:

```
#Const a = 2 + 5

#If a = 7 Then
    Public Sub Alive()
        #If True Then
            Debug.Print 2
        #ElseIf 1 = 2 Then
            Debug.Print 4
        #End If
    End Sub
#Else
    Public Sub Dead()
        Debug.Print 3
    End Sub
#End If
";
```

Into this:

```



    Public Sub Alive()

            Debug.Print 2



    End Sub






```

Thoughts?

**Comment by ThunderFrame** [comment]
Looks good. Does it handle #If/#Else/#ElseIf conditions that include Built-in #Const values like VBA6/VBA7? And conditions with And/Or conditions, and functions/expressions?

**Comment by retailcoder** [comment]
OMG it's done already??!!

**Comment by ThunderFrame** [comment]
Eg. 
```vb
#Const MyConst = 31000
#Const MyOther = 32000
#If not VBA7 And (MyConst >= MyOther Or IsDate(MyConst))
```

**Comment by retailcoder** [comment]
@ThunderFrame Yes, see [VBAPredefinedCompilationConstants.cs](https://github.com/autoboosh/Rubberduck/commit/651dce813cf3e1f939c5fdc2fd10ebdfa99e88be#diff-de6073c35a6de79a40eab55a635c0b91R7) - as for expressions, I would deem them handled, although I couldn't find a test in the [test suite](https://github.com/autoboosh/Rubberduck/commit/651dce813cf3e1f939c5fdc2fd10ebdfa99e88be#diff-04acb0cffc6709ae7429cbdea8dd4fffR10) that combines multiple operators, there's a test harness for every operator you could think of.

@autoboosh awesome work!! did I miss a test for operator precedence?

**Comment by autoboosh** [comment]
The preprocessor should more or less work, yes :)

Here's a quick overview of the design:
1. `VBAPreprocessor `represents a facade into the preprocessing "subsystem"
2. `VBAPreprocessorVisitor `is the actual preprocessor, returning as a result the preprocessed module
3. `VBAExpressionEvaluator `is responsible for the evaluation of VBA conditional compilation constant expressions (a subset of all possible VBA expressions)
4. `VBAPreprocessorVisitor `uses `VBAExpressionEvaluator `to decide what code is "live" or "dead"
5. ILetCoercion represents the implicit coercions possible in VBA (e.g. things like `True ^ #01-01-1900#` are actually possible, which I find quite ridiculous. Might be something for Rubberduck V3? VBA Obfuscator based on replacing "sane" values with the one just mentioned?)

And here are the answers to your feedback:
1. Built-In const values are handled, yes. Project-wide constants aren't. Any ideas here? They seem to be missing in the VBAProject class. In what DLL is the VBE GUI stored? Perhaps we can disassemble it and find out where the hell it saves those constants to...
2. "Composite" expressions are handled as well yes, things like And and Or should work. Everything defined in 5.6.16.2 should work. Although there are no tests for it yet. I'll add some. I also find it weird that the `Is ` operator is supported. You're only allowed to use "primitive" types so `Is `never works with them. The only case I found where it works is Nothing Is Nothing which is True I think. Perhaps I'm missing something here. Any ideas?
3. I'll add operator precedence tests. They're missing currently.
4. IsDate is not a supported function (only a small number of functions are supported per VBA spec) and thus should not work (just tried it in Excel - it throws an error). Does it work for you @ThunderFrame? If so, we'll have to add it.

It still needs a lot more testing but I'm hoping that we can use this in v2.0.

If you find anything you'd like to have changed let me know!

**Comment by Hosch250** [comment]
:+1:

**Comment by retailcoder** [comment]
@autoboosh if all that's missing is project-wide `#Const` values, we'll document it as a known limitation and move on - this is **definitely** going into 2.0! :+1: :+1: :+1:

**Comment by autoboosh** [comment]
There are two limitations so far:
* Project-wide constants are ignored
* Option Compare Database is ignored (it is treated as Text instead)

Other than that it should work in most cases. I'm still trying to add more tests.

**Comment by retailcoder** [comment]
@autoboosh I think this is pretty awesome as is - and it solves more problems than it causes; I'll take the PR anytime!

**Comment by Vogel612** [comment]
Fixed and merged with #1170
